mythril_core/src/acpi/madt.rs

```diff
@@ -0,0 +1,335 @@
+        gsi: u32,
+        /// MPS INI Flags.
+        flags: u16,
+    },
+    /// Non-Maskable Interrupt Source Structure.
+    ///
+    /// See `ACPI § 5.2.12.6`.
+    NmiSource {
+        /// MPS INI Flags.
+        flags: u16,
+        /// Global System Interrupt that this NMI will signal.
+        gsi: u32,
+    },
+    /// Local APIC NMI Structure.
+    ///
+    /// See `ACPI § 5.2.12.7`.
+    LocalApicNmi {
+        /// Processor Object ID.
+        acpi_proc_uid: u8,
+        /// MPS INI Flags.
+        flags: u16,
+        /// Local APIC interrupt input LINTn to which NMI is connected.
+        local_apic_lint: u8,
+    },
+    /// Processor Local x2APIC Structure.
+    ///
+    /// See `ACPI § 5.2.12.12`.
+    LocalX2Apic {
+        /// Processor local x2APIC ID.
+        x2apic_id: u32,
+        /// Local APIC Flags.
+        flags: u32,
+        /// Processor Object ID.
+        apic_proc_uid: u32,
+    },
+}
+
+impl Ics {
+    /// Parse the given
+    fn parse<'a>(ty: IcsType, bytes: &'a [u8]) -> Result<Ics> {
+        ty.check_len(bytes.len())?;
+        match ty {
+            IcsType::ProcessorLocalApic => Ok(Ics::LocalApic {
+                apic_uid: bytes[0],
+                apic_id: bytes[1],
+                flags: NativeEndian::read_u32(&bytes[2..6]),
+            }),
+            IcsType::ProcessorLocalX2Apic => Ok(Ics::LocalX2Apic {
+                x2apic_id: NativeEndian::read_u32(&bytes[2..6]),
+                flags: NativeEndian::read_u32(&bytes[6..10]),
+                apic_proc_uid: NativeEndian::read_u32(&bytes[10..14]),
+            }),
+            IcsType::IoApic => {
+                let ioapic_addr = NativeEndian::read_u32(&bytes[2..6]);
+                Ok(Ics::IoApic {
+                    ioapic_id: bytes[0],
+                    ioapic_addr: ioapic_addr as *const u8,
+                    gsi_base: NativeEndian::read_u32(&bytes[6..10]),
+                })
+            }
+            IcsType::InterruptSourceOverride => {
+                Ok(Ics::InterruptSourceOverride {
+                    source: bytes[1],
+                    gsi: NativeEndian::read_u32(&bytes[2..6]),
+                    flags: NativeEndian::read_u16(&bytes[6..8]),
+                })
+            }
+            IcsType::NmiSource => Ok(Ics::NmiSource {
+                flags: NativeEndian::read_u16(&bytes[0..2]),
+                gsi: NativeEndian::read_u32(&bytes[2..6]),
+            }),
+            IcsType::LocalApicNmi => Ok(Ics::LocalApicNmi {
+                acpi_proc_uid: bytes[0],
+                flags: NativeEndian::read_u16(&bytes[1..3]),
+                local_apic_lint: bytes[3],
+            }),
+            _ => Err(Error::NotImplemented(format!(
+                "type=0x{:x} length={} not implemented",
+                ty as u8,
+                bytes.len()
+            ))),
+        }
+    }
+}
+
+/// Multiple APIC Descriptor Table (MADT).
+///
+/// See `ACPI § 5.2.12`.
+pub struct MADT<'a, 'b: 'a> {
```

Comment:
I've not looked at the spec recently, so I don't recall, but does this relationship between the lifetimes exist? Is the memory from the SDT header in some way related to the ics? It seems like the lifetimes should be the same or entirely separate to me.

Reply:
Tough question. Now that I think about it, I'd think they are entirely separate. You're just reading a pointer from the SDT header.

Reply:
Actually this is needed. The ICS buffer directly follows the SDT and since the reference we hold in this structure is just a slice of the table held in the SDT [rsdt.rs:46](https://github.com/mythril-hypervisor/mythril/blob/rsdt/mythril_core/src/acpi/rsdt.rs#L46), we need to ensure that the lifetime of `'b` outlives that of `'a` or we could use `'a` for both the slice and the SDT since they both should have the same lifetime. Ether works, I don't have much of a preference.

Reply:
Decided to simplify this a bit and just use `'a`.